tests/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Properties/DefineConstantsPropertyPageProviderTests.cs

```diff
@@ -0,0 +1,195 @@
+
+            var setResultValue = await provider.OnSetPropertyValueAsync(DefineConstantsValueProvider.DefineConstantsPropertyName, input, defaultProperties);
+
+            Assert.NotNull(setResultValue);
+            Assert.True(ValidEncoding(setResultValue));
+
+            var intendedInput = getDefinedRepeat(input);
+
+            Assert.Equal(expected: EncodeOutputFormat(intendedInput, true), actual: setResultValue);
+
+            var getResultValue = await provider.OnGetUnevaluatedPropertyValueAsync(DefineConstantsValueProvider.DefineConstantsPropertyName, input, defaultProperties);
+            Assert.NotNull(getResultValue);
+            Assert.Equal(expected: EncodeOutputFormat(intendedInput, false), actual: getResultValue);
+        }
+
+
+        private string formatInput(string[] input)
+        {
+            return string.Join(",", input);
+        }
+
+        private string parseValidInput(string[] input)
+        {
+            string valid = "";
+            for (int i = 0; i < input.Length; i++)
+            {
+                if (ValidInput(input[i]))
+                {
+                    valid += input[i] + ",";
+                }
+            }
+            if (valid.Length > 0)
+            {
+                return valid.Remove(valid.Length - 1);                          //remove the trailing comma
+            }
+            else
+            {
+                return valid;
+            }
+        }
+
+        //determines if input is in valid format
+        private bool ValidInput(string input)
+        {
+            var regex = new Regex(@"(\S+?\s*)\=(\S+?\s*)");
+            return regex.IsMatch(input);
+        }
+
+        //determines if encoding is in valid format
+        private bool ValidEncoding(string encoding)
+        {
+            var regex = new Regex(@"(\S+?\s*)\=[\x22](\S+?\s*)[\x22],?");
+            return regex.IsMatch(encoding);
+        }
+
+        // outputFormat true for the format key="value", false for key=value
+        private string EncodeOutputFormat(string input, bool outputFormat)
+        {
+            input += ",";
+            var regex = new Regex(@"(\S+?\s*)\=(\S+?\s*),");
+            string encoded = "";
+            foreach (System.Text.RegularExpressions.Match match in regex.Matches(input))
+            {
+                string key = match.Groups[1].Value.Replace("/", "//").Replace("=", "/=").Replace(",", "/,");
+                string value = match.Groups[2].Value.Replace("/", "//").Replace("=", "/=").Replace(",", "/,");
+                encoded += outputFormat ? key + "=\"" + value + "\"," : key + "=" + value + ",";
+            }
+            return encoded.TrimEnd(',');
+        }
+
+        //parses repeated keys inputted key value pairs
+        private string getDefinedRepeat(string input)
+        {
+            input += ",";
+            List<KeyValuePair<string, string>> uniqueKeyValuePairs = new List<KeyValuePair<string, string>>();
```

Comment:
Looking at how this collection is used, I think you want a `Dictionary<string, string>` instead. Each call to `FindIndex` in the loop below is O(N), and the loop is O(N), so combining them you get O(N^2), which would be bad for large data sets.

Reply:
This is a test though, so I'm not sure it matters much?